source/slang/slang.cpp

```diff
@@ -1965,6 +1974,87 @@ void Module::_addEntryPoint(EntryPoint* entryPoint)
     m_entryPoints.add(entryPoint);
 }
 
+static bool _canExportDeclSymbol(ASTNodeType type)
+{
+    switch (type)
+    {
+        case ASTNodeType::ModuleDecl:
+        case ASTNodeType::EmptyDecl:
+        case ASTNodeType::NamespaceDecl:
+        {
+            return false;
+        }
+        default: break;
+    }
+
+    return true;
+}
+
+static bool _canRecurseExportSymbol(Decl* decl)
+{
+    if (as<FunctionDeclBase>(decl) ||
+        as<ScopeDecl>(decl))
+    {
+        return false;
+    }
+    return true;
+}
+
+void Module::_processFindDeclsExportSymbolsRec(Decl* decl)
+{
+    if (_canExportDeclSymbol(decl->astNodeType))
+    {
+        // It's a reference to a declaration in another module, so first get the symbol name. 
+        String mangledName = getMangledName(getASTBuilder(), decl);
+
+        Index index = Index(m_mangledExportPool.add(mangledName));
```

Comment:
It's not relevant right now, but I want to keep bringing up that eventually we will probably want to store any large tables of mangled names in a compressed for that collapses out common prefixes.

The reason to collapse out common prefixes is that all the child declarations of some parent type declaration `SomeModule.SomeType` are going to have names of the form `SomeModule.SomeType.*`, so that storing the substring `SomeModule.SomeType` (or rather, the mangling of it) redundantly will get costly.

There are things we can do in the mangling scheme itself to make things more compact, but those are in some ways orthogonal to redundancy across names. The main thing we will want from our mangling scheme when we get aroudn to doing it right is a guarantee that the similarity or "closeness" of two entities is reflected in their mangled names sharing a common prefix. (As a concrete example, we probably *don't* want all mangled function names to start with `F` and all mangled type names to start with `T`, because that gets rid of potential shared prefixes).

I'll need to look up the references on this stuff, but once you have a list of names that you'd like to build into a string pool/table, you can then sort them, and for each entry in the sorted list compute the number N of prefix bytes it shares with its immediate predecessor in the list, and the suffix S that is needs to append to those bytes. You can also compute the index of the earliest entry in the list with which it shares the given prefix. You can then transform each string into a form like

```
struct Entry
{
    size_t prefixBytesSharedWithPrevious;
    size_t indexOfEarliestPredecessorWithMatchingPrefix;
    size_t suffixBytesToAdd;
    char const* suffixBytes;
};
```

Expanding out entries like that into full strings is pretty easy in a linear scan, but more importantly it is also possible to compare a string against an entry in the table without ever fully expanding all the strings. Given an index in the table and a string to test, you can check whether it ends with the given `suffixBytes` for the entry, and then jump to the `indexOfEarliestPredecessorWithMatchingPrefix` and continue matching (with a little bit of careful indexing to make sure you only compare against the bytes that make up the prefix for the chosen entry). IIRC the check should be linear in the length of the string.

A compressed table like that could be used as a way to confirm hit-vs-miss for lookups into a serialized hash table of mangled named in a module.